File: skills/xianyu_compare.js
/**
 * 闲鱼黑胶卖家对比技能 - JavaScript版本
 * 配合 Playwright MCP 使用
 */

// ============ 配置 ============
const CONFIG = {
  seller1: {
    id: '2219735146783',
    name: '音乐大同'
  },
  seller2: {
    id: '1059107164',
    name: '梦的采摘员'
  },
  baseUrl: 'https://www.goofish.com'
};

// ============ 标题标准化 ============
function normalizeTitle(title) {
  return title.toLowerCase()
    .replace(/[·•:：,，、""''「」『』【】《（）\(\)]/g, ' ')
    .replace(/\s+/g, ' ')
    .replace(/黑胶|唱片|专辑|新专辑|限量|带独立编号|带编|日版|台版|cd|lp|1lp|2lp|双|三/g, '')
    .replace(/彩胶|紫胶|红胶|黄胶|绿胶|金胶|灰胶|蓝胶|白胶|透明胶|动画胶/g, '')
    .replace(/电影原声|买家评价|预定|现货|粉丝更优惠|2人小刀价/g, '')
    .replace(/人气第|热销第|\d+名/g, '')
    .replace(/24小时内发布|48小时内发布|72小时内发布|一周内发布/g, '')
    .trim();
}

// ============ 判断是否相同专辑 ============
function isSameAlbum(title1, title2) {
  const n1 = normalizeTitle(title1);
  const n2 = normalizeTitle(title2);

  if (n1 === n2) return true;
  if (n1.length > 10 && n2.includes(n1)) return true;
  if (n2.length > 10 && n1.includes(n2)) return true;
  return false;
}

// ============ 提取专辑标题（清理版） ============
function cleanAlbumTitle(rawTitle) {
  return rawTitle
    .replace(/¥\s*\d+/g, '')
    .replace(/\d+人想要/g, '')
    .replace(/24小时内发布|48小时内发布|72小时内发布|一周内发布/g, '')
    .replace(/热销第\d+名|人气第\d+名/g, '')
    .replace(/买家评价["'][^"']*["']/g, '')
    .replace(/预定|现货|粉丝更优惠/g, '')
    .replace(/2人小刀价/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

// ============ 抓取卖家数据 ============
async function scrapeSeller(page, sellerId) {
  // 导航到卖家页面
  await page.goto(`${CONFIG.baseUrl}/personal?userId=${sellerId}`);

  // 等待页面加载
  await page.waitForTimeout(2000);

  // 点击"在售"标签
  await page.evaluate(() => {
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT);
    let node;
    while (node = walker.nextNode()) {
      if (node.textContent === '在售') {
        let parent = node.parentElement;
        while (parent && parent !== document.body) {
          if (parent.classList.contains('tab') || parent.getAttribute('role') === 'tab') {
            parent.click();
            return;
          }
          parent = parent.parentElement;
        }
        if (node.parentElement) node.parentElement.click();
        break;
      }
    }
  });

  await page.waitForTimeout(2000);

  // 滚动加载所有商品
  await page.evaluate(async () => {
    for (let round = 0; round < 15; round++) {
      for (let i = 0; i < 10; i++) {
        window.scrollBy(0, 500);
        await new Promise(r => setTimeout(r, 100));
      }
      await new Promise(r => setTimeout(r, 500));
    }
  });

  // 提取商品数据
  return await page.evaluate(() => {
    const seen = new Set();
    const products = [];

    document.querySelectorAll('a[href*="item?id="]').forEach(link => {
      const id = link.href.match(/id=(\d+)/)?.[1];
      if (!id || seen.has(id)) return;
      seen.add(id);

      let text = (link.innerText || '').trim();
      if (text.length < 5) return;

      let title = text
        .replace(/¥\s*\d+/g, '')
        .replace(/\d+人想要/g, '')
        .replace(/24小时内发布|48小时内发布|72小时内发布|一周内发布/g, '')
        .replace(/热销第\d+名|人气第\d+名/g, '')
        .replace(/买家评价["'][^"']*["']/g, '')
        .replace(/预定|现货|粉丝更优惠/g, '')
        .replace(/2人小刀价/g, '')
        .replace(/\s+/g, ' ')
        .trim();

      if (title.length > 5) {
        products.push({ id, title });
      }
    });

    return products;
  });
}

// ============ 对比数据 ============
function compareSellers(seller1Data, seller2Data) {
  const processedS1 = new Set();
  const processedS2 = new Set();

  const overlapping = [];
  const s1Only = [];
  const s2Only = [];

  // 找出重叠和seller1独有
  for (const p1 of seller1Data) {
    if (processedS1.has(p1.title)) continue;
    processedS1.add(p1.title);

    let found = false;
    for (const p2 of seller2Data) {
      if (processedS2.has(p2.title)) continue;
      if (isSameAlbum(p1.title, p2.title)) {
        overlapping.push(p1.title);
        processedS2.add(p2.title);
        found = true;
        break;
      }
    }
    if (!found) {
      s1Only.push(p1.title);
    }
  }

  // 找出seller2独有
  for (const p2 of seller2Data) {
    if (!processedS2.has(p2.title)) {
      s2Only.push(p2.title);
    }
  }

  return { overlapping, seller1Only: s1Only, seller2Only: s2Only };
}

// ============ 打印报告 ============
function printReport(result, seller1Name, seller2Name) {
  console.log('\n' + '='.repeat(40));
  console.log('   闲鱼卖家黑胶唱片对比');
  console.log('='.repeat(40) + '\n');

  console.log('【统计】');
  console.log(`  ${seller1Name}: ${result.seller1Only.length + result.overlapping.length} 张`);
  console.log(`  ${seller2Name}: ${result.seller2Only.length + result.overlapping.length} 张`);
  console.log(`  重叠: ${result.overlapping.length} 张`);
  console.log(`  ${seller1Name}独有: ${result.seller1Only.length} 张`);
  console.log(`  ${seller2Name}独有: ${result.seller2Only.length} 张`);

  console.log(`\n【重叠专辑】(${result.overlapping.length}张)`);
  result.overlapping.forEach((album, i) => {
    console.log(`  ${i+1}. ${album.substring(0, 55)}`);
  });

  console.log(`\n【${seller1Name}独有专辑】(${result.seller1Only.length}张)`);
  result.seller1Only.forEach((album, i) => {
    console.log(`  ${i+1}. ${album.substring(0, 55)}`);
  });

  console.log(`\n【${seller2Name}独有专辑】(${result.seller2Only.length}张)`);
  result.seller2Only.forEach((album, i) => {
    console.log(`  ${i+1}. ${album.substring(0, 55)}`);
  });

  console.log('\n' + '='.repeat(40) + '\n');
}

// ============ 导出接口 ============
module.exports = {
  CONFIG,
  normalizeTitle,
  isSameAlbum,
  cleanAlbumTitle,
  scrapeSeller,
  compareSellers,
  printReport
};

File: skills/xianyu_compare.py
#!/usr/bin/env python3
"""
闲鱼黑胶卖家对比技能
Xianyu Vinyl Sellers Comparison Skill

用法:
    python xianyu_compare.py                    # 使用默认卖家
    python xianyu_compare.py --seller1 xxx --seller2 yyy
"""

import json
import re
import sys
from datetime import datetime
from pathlib import Path

# 如果安装了 playwright，可以使用：
# from playwright.sync_api import sync_playwright

# Default seller IDs
SELLER1_ID = "2219735146783"  # 音乐大同
SELLER2_ID = "1059107164"     # 梦的采摘员
OUTPUT_DIR = Path("./output")


def normalize_title(title: str) -> str:
    """标准化专辑标题，去除干扰信息"""
    title = title.lower()

    # 移除特殊字符
    for c in '·•:：,，、""''「」『』【】《（）()':
        title = title.replace(c, ' ')

    # 移除多余空格
    title = re.sub(r'\s+', ' ', title)

    # 移除常见后缀
    keywords = [
        '黑胶', '唱片', '专辑', '新专辑', '限量', '带独立编号', '带编',
        '日版', '台版', 'cd', 'lp', '1lp', '2lp', '双', '三',
        '彩胶', '紫胶', '红胶', '黄胶', '绿胶', '金胶', '灰胶', '蓝胶',
        '白胶', '透明胶', '动画胶', '电影原声', '买家评价', '预定',
        '现货', '粉丝更优惠', '2人小刀价', '人气第', '热销第',
        '24小时内发布', '48小时内发布', '72小时内发布', '一周内发布'
    ]
    for kw in keywords:
        title = title.replace(kw, '')

    return title.strip()


def is_same_album(title1: str, title2: str) -> bool:
    """判断两个专辑名是否相同"""
    n1 = normalize_title(title1)
    n2 = normalize_title(title2)

    if n1 == n2:
        return True
    if len(n1) > 10 and n1 in n2:
        return True
    if len(n2) > 10 and n2 in n1:
        return True
    return False


def compare_sellers(seller1_data: list, seller2_data: list) -> dict:
    """对比两个卖家的专辑"""
    processed_s1 = set()
    processed_s2 = set()

    overlapping = []
    s1_only = []
    s2_only = []

    # 找出重叠和seller1独有
    for album1 in seller1_data:
        if album1 in processed_s1:
            continue
        processed_s1.add(album1)

        found = False
        for album2 in seller2_data:
            if album2 in processed_s2:
                continue
            if is_same_album(album1, album2):
                overlapping.append(album1)
                processed_s2.add(album2)
                found = True
                break

        if not found:
            s1_only.append(album1)

    # 找出seller2独有
    for album2 in seller2_data:
        if album2 not in processed_s2:
            s2_only.append(album2)

    return {
        "overlapping": overlapping,
        "seller1_only": s1_only,
        "seller2_only": s2_only
    }


def print_report(result: dict, seller1_name: str, seller2_name: str):
    """打印对比报告"""
    print('\n' + '=' * 40)
    print('   闲鱼卖家黑胶唱片对比')
    print('=' * 40 + '\n')

    print('【统计】')
    print(f'  {seller1_name}: {len(result["seller1_only"]) + len(result["overlapping"])} 张')
    print(f'  {seller2_name}: {len(result["seller2_only"]) + len(result["overlapping"])} 张')
    print(f'  重叠: {len(result["overlapping"])} 张')
    print(f'  {seller1_name}独有: {len(result["seller1_only"])} 张')
    print(f'  {seller2_name}独有: {len(result["seller2_only"])} 张')

    print(f'\n【重叠专辑】({len(result["overlapping"])}张)')
    for i, album in enumerate(result["overlapping"], 1):
        print(f'  {i}. {album[:55]}')

    print(f'\n【{seller1_name}独有专辑】({len(result["seller1_only"])}张)')
    for i, album in enumerate(result["seller1_only"], 1):
        print(f'  {i}. {album[:55]}')

    print(f'\n【{seller2_name}独有专辑】({len(result["seller2_only"])}张)')
    for i, album in enumerate(result["seller2_only"], 1):
        print(f'  {i}. {album[:55]}')

    print('\n' + '=' * 40 + '\n')


def main():
    """主函数"""
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # 读取数据文件（假设已由浏览器爬取生成）
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')

    # 这里应该读取爬取的数据
    # 示例：seller1_file = OUTPUT_DIR / f"seller1_{timestamp}.json"

    print("请先使用浏览器爬取两个卖家的数据，然后运行对比分析")
    print("\n使用方法:")
    print("1. 打开闲鱼网页并登录")
    print("2. 访问卖家页面并爬取数据")
    print("3. 保存为JSON文件")
    print("4. 运行此脚本进行对比")


if __name__ == "__main__":
    main()
